Improve runtime by concatenation of opcodes.

fbmrk opened the pull request:
I made a statistic about vm opcodes. If you are running `jerryscript` with the `tests/jerry/*.js` files the opcode after the `VM_OC_LESS` will be the `VM_OC_BRANCH_IF_TRUE` in 95.5%. And this value for sunspider js test files is 90%. This is mostly caused by the `for` statement. You often use it like this:
`for ( i = 0; i `**<**` x; i++ ) { ... }`

Taking advantage of this we can improve the runtime.

Benchmark | Perf (sec) |
----: | ----: | 
3d-cube.js | 0.859 -> 0.837 : +2.623% | 
3d-raytrace.js | 1.067 -> 1.065 : +0.194% | 
access-binary-trees.js | 0.572 -> 0.576 : -0.694% | 
access-fannkuch.js | 2.221 -> 2.118 : +4.654% | 
access-nbody.js | 1.156 -> 1.142 : +1.166% | 
bitops-3bit-bits-in-byte.js | 0.594 -> 0.590 : +0.757% | 
bitops-bits-in-byte.js | 0.879 -> 0.799 : +9.059% | 
bitops-bitwise-and.js | 1.042 -> 1.012 : +2.927% | 
bitops-nsieve-bits.js | 1.376 -> 1.321 : +3.997% | 
controlflow-recursive.js | 0.389 -> 0.402 : -3.246% | 
crypto-aes.js | 0.963 -> 0.936 : +2.806% | 
crypto-md5.js | 0.647 -> 0.647 : +0.007% | 
crypto-sha1.js | 0.637 -> 0.641 : -0.730% | 
date-format-tofte.js | 0.753 -> 0.739 : +1.804% | 
date-format-xparb.js | 0.406 -> 0.403 : +0.657% | 
math-cordic.js | 1.284 -> 1.256 : +2.227% | 
math-partial-sums.js | 0.763 -> 0.775 : -1.574% | 
math-spectral-norm.js | 0.576 -> 0.561 : +2.606% | 
string-base64.js | 1.517 -> 1.520 : -0.217% | 
string-fasta.js | 1.288 -> 1.285 : +0.258% | 
Geometric mean: | +1.498% | 

Binary sizes (bytes)
d7c710d3d4:137584
a254abe9f1:137672

JerryScript-DCO-1.0-Signed-off-by: Marko Fabo mfabo@inf.u-szeged.hu

akosthekiss commented on jerry-core/vm/vm.c:
```diff
@@ -2010,10 +2010,49 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
+                default:
+                {
+                  JERRY_ASSERT (CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p) == 2
+                                || CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p) == 3);
```
How can this happen? `case 3` has already been handled above.

akosthekiss replied:
Ahh, FALLTHROUGH. That's evil :)

akosthekiss commented on jerry-core/vm/vm.c:
```diff
@@ -2010,10 +2010,49 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
+                  branch_offset |= *(byte_code_p++);
+                  break;
+                }
+              }
```
OK, let's try another question. So, this is basically a loop unrolling. Have you tried it this way? (Seems a bit more logical to me.)

```c
JERRY_ASSERT (CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p) >= 1
              || CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p) <= 3);
switch (CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p))
{
  case 3:
  {
    branch_offset = *(byte_code_p++);
    branch_offset <<= 8;
    /* FALLTHRU */
  }
  case 2:
  {
    branch_offset |= *(byte_code_p++);
    branch_offset <<= 8;
    /* FALLTHRU */
  }
  default:
  {
    branch_offset |= *(byte_code_p++);
    break;
  }
}
```

fbmrk replied:
I haven't tried it this way. I followed this part of the code when I wrote my patch:
https://github.com/jerryscript-project/jerryscript/blob/master/jerry-core/vm/vm.c#L852
But yes, it seems more logical. I will test it. 
Should I change the referenced code too?

akosthekiss replied:
I think that @zherczeg should rather comment on this.

zherczeg replied:
The 1 byte argument has the highest chance. If you want to change it try something like:

```
branch_offset = *(byte_code_p++);
if (unlikely (CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p) > 1)) {
   handle case 2 and 3.
}
```

akosthekiss replied:
This could be the most compact (and potentially also the fastest) form then:

```c
branch_offset_length = CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p);
JERRY_ASSERT (branch_offset_length >= 1 && branch_offset_length <= 3);
branch_offset = *(byte_code_p++);
if (unlikely (branch_offset_length != 1))
{
  branch_offset <<= 8;
  branch_offset |= *(byte_code_p++);
  if (unlikely (branch_offset_length == 3))
  {
    branch_offset <<= 8;
    branch_offset |= *(byte_code_p++);
  }
}
```
@zherczeg opinion?

akosthekiss commented on jerry-core/vm/vm.c:
```diff
@@ -2010,10 +2010,49 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
+                  JERRY_ASSERT (CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p) == 2
+                                || CBC_BRANCH_OFFSET_LENGTH (*byte_code_start_p) == 3);
+
+                  branch_offset <<= 8;
```
If this approach is kept, then this statement can be hoisted to `case 3` above. (If `case 2`, there is no point in shifting 0 left by 8 bits.)

fbmrk commented:
@akosthekiss @zherczeg 
I've updated the PR.
NOTE: I've also changed this:
```
branch_offset = -branch_offset;
...
byte_code_p = byte_code_start_p + branch_offset;
```
-->
```
byte_code_p = byte_code_start_p - branch_offset;
```

I'll update the table too.

fbmrk commented:
Benchmark | Perf (sec) |
----: | ----: | 
3d-cube.js | 0.859 -> 0.838 : +2.516% | 
3d-raytrace.js | 1.067 -> 1.065 : +0.221% | 
access-binary-trees.js | 0.572 -> 0.573 : -0.115% | 
access-fannkuch.js | 2.221 -> 2.137 : +3.794% | 
access-nbody.js | 1.156 -> 1.141 : +1.292% | 
bitops-3bit-bits-in-byte.js | 0.594 -> 0.584 : +1.698% | 
bitops-bits-in-byte.js | 0.879 -> 0.804 : +8.483% | 
bitops-bitwise-and.js | 1.042 -> 0.984 : +5.566% | 
bitops-nsieve-bits.js | 1.376 -> 1.308 : +4.919% | 
controlflow-recursive.js | 0.389 -> 0.396 : -1.602% | 
crypto-aes.js | 0.963 -> 0.928 : +3.573% | 
crypto-md5.js | 0.647 -> 0.642 : +0.651% | 
crypto-sha1.js | 0.637 -> 0.633 : +0.499% | 
date-format-tofte.js | 0.753 -> 0.733 : +2.595% | 
date-format-xparb.js | 0.406 -> 0.404 : +0.493% | 
math-cordic.js | 1.284 -> 1.255 : +2.264% | 
math-partial-sums.js | 0.763 -> 0.766 : -0.467% | 
math-spectral-norm.js | 0.576 -> 0.560 : +2.657% | 
string-base64.js | 1.517 -> 1.521 : -0.286% | 
string-fasta.js | 1.288 -> 1.276 : +0.979% | 
Geometric mean: | +2.015% | 

Binary sizes (bytes)
d7c710d3d4:137584
ee60ff634f:137680

akosthekiss commented on jerry-core/vm/vm.c:
```diff
@@ -2010,10 +2001,37 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
+
+              if (is_less)
+              {
+                /* Note: The opcode is a bacward branch. */
```
typo: backward

akosthekiss commented on jerry-core/vm/vm.c:
```diff
@@ -2010,10 +2001,37 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
+              if (is_less)
+              {
+                /* Note: The opcode is a bacward branch. */
+                byte_code_p = byte_code_start_p - branch_offset;
```
btw, if `is_less` is false, did we do the `branch_offset` computation above in vain? if `!is_less`, wouldn't be `byte_code_p += branch_offset_length` enough?

fbmrk replied:
yes, thanks! updated.

akosthekiss commented on jerry-core/vm/vm.c:
```diff
@@ -2010,10 +2001,37 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
-            ecma_integer_value_t right_integer = (ecma_integer_value_t) right_value;
+            bool is_less = (ecma_integer_value_t) left_value < (ecma_integer_value_t) right_value;
+
+            if (*byte_code_p <= CBC_BRANCH_IF_TRUE_BACKWARD_3 && *byte_code_p >= CBC_BRANCH_IF_TRUE_BACKWARD)
```
BTW, might be worth adding a note here that this is actually a lookahead to the next opcode and executing a CBC_BRANCH_IF_TRUE_BACKWARD* right here if that's what is found (I'm not sure that such lookaheads are used anywhere else, so it might be good mentioning.)

fbmrk commented:
Done.
